### ISSUE TITLE
Shipping calculator (aka estimator) is unfinished.

### ISSUE THREAD

**Issue by PloughGuy** [cubecart/v6#1882]
Existing capabilities are fundamentally good. There are some wrinkles with the Workflow however. 

# The Short Version
The View Cart page needs some simple overhauls to present the shipping calculator (it is not an estimator - see below) at its full magnificence.  

All of these suggestions relate to the View Cart page, not the Check Out page unless stated otherwise.

- non-logged-in customers should always go to the Check-out page via the View Cart page for a clear and uncluttered view of their cart and ready access to the shipping calculator before they are confronted by the Checkout page.
- In the View Cart page, the shipping options list box should be concealed until a customer-provided country and postcode are available.  This means the shipping price box will be empty, so...
- If the shipping price box is empty, replace it with a "Get shipping options" link (one that _looks_ like a link.)  Or perhaps "Calculate shipping and taxes".
- Shipping prices should always be accompanied by the destination such as "Shipping:  |   $9.95| to AU 0650" beside or under the selected price in the totals column or "Shipping options for delivery to Australia postcode 2048" above the shipping options list box.  This will minimise customer confusion and consequent shopkeeper grief, expense and lost time.  This applies to the Checkout page also.
- if a shipping option has been selected, there should be a "Change postcode" link nearby that re-opens the shipping calculator data entry window.
- there should be a setting to remove the region from the calculator interface.  This is only required for jurisdictions that have regional taxes, and it is an unnecessary imposition for jurisdictions that do not.

# Motivation
**First point:**  _This is not an estimate from the customer’s viewpoint_. This is the price charged if the cart is not changed.  Therefore this is a shipping calculator,  not an estimator.  Its accuracy is only of concern to the shopkeeper, who has the responsibility of ensuring that dimensions and weights fed to the shipping modules are accurate, and that appropriate shipping products are selected.  And since there is no way to feed dimensions to the shipping modules, the shopkeeper is fundamentally stuffed until someone fixes it.  But this is not the customer's concern.  The price is the price.

**Second point.**  _We should not quote postage unless we know the destination_.  This means the customer has logged in and we have their account postcode, or they have provided a postcode to the shipping calculator.  The customer currently sees the lowest possible price (to the postcode of the store) followed by a statement "This is an estimate."  I myself did not realise that "Estimate" was a verb here, and I did not perceive it as a link.  I thought the blue text was for emphasis.  Therefore I did not realise there was a shipping calculator build into the store until the the Beautifully Bearded Bsmither pointed it out.  American customers will be shooting holes in their screens when they enter their details and the shipping price goes up by a factor of three.  Australian customers will vomit beer with pie fragments over their iPads.  English customers will say "Oh, I say...  What?  By jove!" or maybe "Egad!"  

**Third point.** _We must list the destination country and postcode with the shipping price_.
This makes the shipping price unambiguous to the customer.

**Fourth point**  _The postcode defines the region_, therefore we should not ask for it.  We do not need it to calculate shipping.  Some jurisdictions may need it to calculate tax.  At the very least, there should be a configuration setting to hide it for jurisdictions that do not have regional taxes.  (In the future, CubeCart 12.6 will determine its necessity automatically based on tax rules.  When the region is required, it should be derived from some appropriate service such as geonames, which is free, or one of the faster paid address-validators.)

The procedure should be:
* Cart icon pop-up: Users who are not not logged in always go to the View Cart page rather than the checkout page.  This gives them an uncluttered view of their order and a clear invitation to confirm shipping costs;
* Cart icon pop-up: Perhaps non-logged-in users might see a “View Cart and Check Out” button rather than two separate buttons.
* View Cart: Do not display the shipping option prices list box if there is no customer-provided postcode;
* View Cart: If there is no customer-provided postcode, replace the shipping price box with a ‘Calculate shipping’ link to the shipping calculator i.e. instead of a price there is a link to the Calculator e.g.  "Shipping:    _Calculate shipping price_". 
* Check-out: If there is no customer-provided postcode, replace the shipping options list box with "Complete postcode and country above for shipping options" or provide this as the only entry in the list box.
* View Cart and Check-out: Add country and postcode to the caption above the shipping options list box when shipping options are available.
* View Cart and Check-out: Add country and postcode after or under the shipping price in the totals column.
* Everywhere: Do not, in any customer-visible page, use the word "Estimate".  (a) it is a verb and a noun, so it's meaning is unclear without further context, and (b) it is only an estimate for the vendor - customers expect to select a firm shipping product from the options offered, pay the money and receive the goods with no further negotiation about the price.
* View Cart: if a shipping price is displayed, it has a “change postcode“ link under the postcode.

Most of these changes are cosmetic and can be implemented by tweaking the templates.  However, this should not be required for a new store.  It should be correct out of the box.

**Comment by abrookbanks** [comment]
Thanks for the feedback there is some good stuff here. Closed as a feature request as per the [contributing docs](https://github.com/cubecart/v6/blob/6.1.14-branch/CONTRIBUTING.md).

**Comment by abrookbanks** [comment]
Something for version 7 I think. There are some significant core complexities around this demanding ground up rewrite.

**Comment by PloughGuy** [comment]
Demanding rewrite indeed!  Says the man who sorted out the sequential order number problem in a matter of minutes.
Gotta say, I looked at it, thought "mostly templates, coupla hours tops".  But then I didn't think about it much.  Never do.  Well, I did chew an HB down to the point thinking about the customer side of it.  Tasty bugger it was, too.  Lakeland, of course...

**Comment by abrookbanks** [comment]
Haha. Thanks for this. I really like it all but its really is deeply complex to change. We could probably cherry pick a few bits but in order to do this properly we need to start over.

**Comment by PloughGuy** [comment]
Any easy bits that can make it into 6.2 would be greatly appreciated.  It will mean I will have less divergence.  Even just the cosmetic changes would be enough.

Something else for 6.2 is a standard interface for address validation modules.  I currently use geonames to confirm country, state, postcode and town (which has been fine performance wise) but there are others.) I injected it into 4.3.4 in about 2011 and returns completely stopped.

People need a choice in these things, so an address validator interface with geonames as the default module would be an ideal solution. If you do the interface, I will write the module for you to take over and Al-ify to your hearts content.

Oh, and another thing that would be nice for 6.2: Incorporate packed dimensions into the entire solution - add fields for packed weight and three dimensions to order summary, update the shipping modules so they can use dimensions for shipping quotes, and provide a formal interface for a packing module.  Packing algorithms are generally PhD projects for advanced applied mathematicians, but many ranges of products lend themselves to simple algorithms.  A nice $100 plugin could take the dimensions from the product information and rumble around in the CPU for 100ms and find the smallest box of the available options to pack it in.  A tricker problem is to fit into the variable dimensions of a satchel.  

The packer has to allow for packing materials, so products may need a packing rule ("Needs bubblewrap") that is understood by the module and used to adjust the budget for filler space.

I have written one for my store that is about 98% correct.  We use actual postage always, no free, percentage or fixed price postage, and it is generally spot on.  We sell notebooks (paper ones from Belgium) and they fit nicely into a large letter at$.5.50 anywhere in Australia, provided we take about 6 pages out to keep it under 500 grams packed.  For example, we have a packing rule for A4 notebooks that can be shipped this way called "Page sacrifice".  This is offered to customers as an option along with parcel post at up to double the price. This rule is not necessary for A5 notebooks because a large letter will always hold one of them.  There is another named Punch that ensures we don't try to put a punch in a satchel because we have had problems in the past.  Packing rules are set by an additional field in inventory maintenance.

My shipping calculator chooses the optimal solution from letters, parcels, and several sizes of satchel.  This means my postage actuals are very close to my postage billed, which helps my meagre bottom line immensely.

One of my big projects now is to integrate this masterpiece into 6.1.

I don't think CubeCart, at its bargain price of $FREE, should provide a packing module.  But I do think it should provide a defined interface so Havenswift, Noodleman and other quality vendors of fine enhancements can provide a product to suit (and get their PhDs at the same time.)  it will also make CC more attractive to advanced shopkeepers if they can properly manage shipping costs.

The components (as I see them) are:
1: Add fields to order_summary to hold the packing algorithm results:
  packed weight
  packed height
  packed width
  packed length

I also have a text field "packing_solution" that holds a log from the packer so I can analyse any weird solutions.  This was fundamental to getting a working, useful analyser.  I would add that too.  It could be logged but it is easier to find if it is part of order sum.  I added a "Packing analysis" button in the order admin to display it.

Add a hook in the cart before the shipping options are generated to call the packer.

Add some code to the existing shipping modules to check for dimensions in order_sum and use them rather than the standard dimensions in the module settings.  I expect the same bit of code would work for all shipping modules.
